renku/core/session/notebook_service.py

```diff
@@ -0,0 +1,386 @@
+
+    def _get_in_progress_gitlab_job_id(self, gitlab_project_name: str, commit_sha: str) -> Optional[str]:
+        """Get the ID of any job that is in progress and matches the specified arguments."""
+        project_name_enc = gitlab_project_name.replace("/", "%2F")
+        res = self._send_renku_request(
+            "get",
+            f"{self._renku_url()}/api/projects/{project_name_enc}/jobs",
+            params={"scope": ["running", "pending", "created"]},
+            headers=self._auth_header(),
+        )
+        if res.status_code == 200:
+            for job in res.json():
+                if job.get("commit", {}).get("id") == commit_sha:
+                    return job["id"]
+        return None
+
+    def _wait_for_ci_jobs_completion(
+        self,
+        gitlab_project_name: str,
+        commit_sha: str,
+    ):
+        """Wait for all in-progress CI jobs related to the arguments to finish."""
+        if not self._is_user_registered():
+            return
+        now = datetime.now()
+        timeout_time = now + timedelta(seconds=self.DEFAULT_TIMEOUT_SECONDS)
+        while now < timeout_time:
+            if self._get_in_progress_gitlab_job_id(gitlab_project_name, commit_sha) is None:
+                return
+            sleep(5)
+        raise errors.NotebookServiceSessionError(
+            f"Waiting for the CI jobs for project {gitlab_project_name} to finish timed out."
+        )
+
+    def _commit_and_push_checks(self):
+        client = _get_client()
+        remote = get_remote(client.repository)
+        if self._is_user_registered():
+            if client.repository.is_dirty() or len(client.repository.untracked_files) > 0:
+                communication.confirm(
+                    "You have new uncommitted or untracked changes to your repository. "
+                    "Renku can automatically commit and push these changes so that it builds "
+                    "the correct image for your session. Do you wish to proceed?",
+                    abort=True,
+                )
+                client.repository.add(all=True)
+                client.repository.commit("Automated commit by Renku CLI.")
+                client.repository.push()
+            remote_head_hexsha = client.repository.run_git_command(
+                "rev-parse", f"{remote.name}/{client.repository.active_branch}"
+            )
+            if client.repository.head.commit.hexsha != remote_head_hexsha:
+                communication.confirm(
+                    "You have new changes that are present only in your local repository. "
+                    "Renku can automatically push these changes so that it builds "
+                    "the correct image for your session. Do you wish to proceed?",
+                    abort=True,
+                )
+                client.repository.push()
+
+    @staticmethod
+    def _send_renku_request(req_type: str, *args, **kwargs):
```

Comment:
We should probably use these methods instead: https://github.com/SwissDataScienceCenter/renku-python/blob/develop/renku/core/util/requests.py#L63 they also have retry logic with exponential backoff

Reply:
@Panaetius _send_renku_request is using `renku.core.session.utils.requests`. The requests in the method is coming from `renku.core.session.utils`